src/sms.rs

```diff
@@ -0,0 +1,201 @@
+    fn encode_number(number: &str) -> Result<String<15>, Error> {
+        let mut encoded_number = String::from(number.trim_start_matches("+"));
+
+        if encoded_number.len() % 2 != 0 {
+            encoded_number
+                .push('F')
+                .map_err(|_| Error::BufferTooSmall(None))?;
+        }
+
+        if number.is_ascii() {
+            // Since we are checking if the number of characters is even before this unsafe
+            // and we only allow ASCII chars doing this swap shouldn't have any UB
+            unsafe {
+                for c in encoded_number.as_bytes_mut().array_chunks_mut::<2>() {
+                    c.swap(0, 1);
+                }
+            }
+
+            Ok(encoded_number)
+        } else {
+            Err(Error::SmsNumberNotAscii)
+        }
+    }
+    // Convert a ASCII string to GSM 7bit
+    // Reimplement from https://github.com/nrfconnect/sdk-nrf/blob/main/lib/sms/string_conversion.c#L162
+    fn ascii_to_gsm7bit<const N: usize>(text: &str) -> Result<String<N>, Error> {
+        let mut encoded_message = String::new();
+
+        for c in text.chars() {
+            if c.is_ascii() {
+                let char_7bit = ASCII_TO_7BIT_TABLE[c as usize];
+                if char_7bit & STR_7BIT_ESCAPE_IND == 0 {
+                    encoded_message
+                        .push(char_7bit as char)
+                        .map_err(|_| Error::BufferTooSmall(None))?;
+                } else {
+                    encoded_message
+                        .push(STR_7BIT_ESCAPE_CODE as char)
+                        .map_err(|_| Error::BufferTooSmall(None))?;
+                    encoded_message
+                        .push((char_7bit & STR_7BIT_CODE_MASK) as char)
+                        .map_err(|_| Error::BufferTooSmall(None))?;
+                }
+            }
+        }
+
+        Ok(encoded_message)
+    }
+    // Pack a GSM 7 bit strings into 7 bites without 1 bit padding
+    // Reimplement from https://github.com/nrfconnect/sdk-nrf/blob/main/lib/sms/string_conversion.c#L294
+    fn pack_gsm7bit<const N: usize>(text: String<N>) -> Vec<u8, N> {
+        let mut src: usize = 0;
+        let mut dst: usize = 0;
+        let mut shift: usize = 0;
+        let len = text.len();
+        let mut bytes = text.into_bytes();
+
+        while src < len {
+            bytes[dst] = bytes[src] >> shift;
+            src += 1;
+            if src < len {
+                bytes[dst] |= bytes[src] << (7 - shift);
+                shift += 1;
+                if shift == 7 {
+                    shift = 0;
+                    src += 1;
+                }
+            }
+            dst += 1;
+        }
+        bytes.truncate(dst);
+        bytes
+    }
+    /// Sends the craftes message
+    /// `N` is need to provide internal buffer size for message and number encoding. Needs to be at least 2 * message.len() + 34
```

Comment:
I wish const generics were further along, then you could do the calculation in code.

Maybe for now, for lazy people, maybe also document the biggest number you could ever need.

Thinking about it, that's just 354 bytes. Honestly, given the performance of the nrf91, it might just be worth it to always use that amount so that the API can be simplified. A 354 byte bigger stack is not the end of the world.

Reply:
Yes I was looking into const generic too, with it we could even split out the size of at_cmgs from ascii_to_gsm7bit buffer and just calculate everything with const generics.
For now I add the documentation about max 354 bytes. I hoped to add a default value of 354 but const_default_generics are not allowed inside function bodies :/.
I agree that is not the end of the world but they can be case where you just sending some few characters length for debugging and using a buffer of over 300 bytes just sounds wasteful and since we have const generics we can do better then C 😅

Reply:
Sure, then let's keep it. With the documentation as it is now I think it's fine